EIPS/eip-5283.md

```diff
@@ -0,0 +1,107 @@
+
+
+## Rationale
+
+The only available RPG construction today is based on using a contract storage cell. This construction is clean but it is not forward-compatible with transaction execution parallelization.
+
+Several EVM-based blockchains have successfully tested designs for the parallelization of the EVM (i.e. see [RSKIP-144](https://github.com/rsksmart/RSKIPs/blob/master/IPs/RSKIP144.md)). The best results have been obtained with fine-grained parallelization where conflicts are detected by tracking writes and reads of individual storage cells. The designs based on tracking the use of accounts or contracts provide only minor benefits, because most transactions use the same ERC20 contracts.
+
+The address `0x0A` is the next one available within the rage defined by [EIP-1352](https://eips.ethereum.org/EIPS/eip-1352).
+
+### Sample usage
+
+```
+pragma solidity ^0.8.0;
+
+abstract contract ReentrancyGuard2 {
+
+    uint8 constant SemaphoreAddress = 0x0A;
+    /**
+     * @dev Prevents a contract from calling itself, directly or indirectly.
+     * Calling a `nonReentrant` function from another `nonReentrant`
+     * function is supported.      
+     */
+    modifier nonReentrant() {
+        _nonReentrantBefore();
+        _;
+    }
+
+    function _nonReentrantBefore() private {
+    	assembly {
+            if iszero(staticcall(1000,SemaphoreAddress,  0, 0, 0, 0)) {
+                revert(0, 0)
+            }
+        }
+    }
+}
+```
+
+
+### Parallellizable storage-based RPGs
+ 
+The only way to paralellize prexistent contracts that are using the storage RPG construction is that the VM automatically detects that a storage variable is used for the RPG, and proves that is works as required. This requires static code analysys. This is difficult to implement in consensus for two reasons. First, the CPU cost of detection and/or proving may be high. Second, some contract functions may not be protected by the RPG, meaning that some execution paths do not alter the RPG, which may complicate proving. Therefore this proposal aims to protect future contracts and let them be parallelizable, rather than to paralellize already deployed ones.
+
+### Alternatives
+
+There are alternative designs to implement RPGs on the EVM:
+
+1.	[EIP-1153: Transient storage opcodes](https://eips.ethereum.org/EIPS/eip-1153) or [`SSTORE_COUNT`](https://github.com/ethereum/EIPs/issues/119). The first does reduce the cost of an RPG, but does not address the problem of parallelization. The second enables execution parallelization but it is way more complex as it returns the number `SSTORE` opcodes executed, not the number of reentrant calls. Reentrancy must be deducted from this value. Also the `SSTORE` count is not currently tracked by EVM. Also, implementing new opcodes is discuraged if the same functionality can be implemented using precompiles.
```

Comment:
Instead of linking to these, you should briefly describe their behavior.  This EIP won't be able to move to final until all of the EIPs it references are also final, and since this EIP is "competing" with these you certainly wouldn't want to be blocked by them not being included.

Reply:
Ok, I'll describe them. But I can still mention the IP numbers right ?

Reply:
Not unless you want to be blocked by them.  Until an EIP is final, it is completely mutable.  If you refer to EIP-X today and then your EIP becomes final but EIP-X is not, then it is possible that EIP-X will be radically changed before it is finally merged as final and now your EIP doesn't make sense because it is referring to a document that is not the same as when you wrote the reference.

Reply:
Understood!